### ISSUE TITLE
Dropout

### ISSUE THREAD

**Issue by harmdevries89** [bartvm/blocks#280]
I would like to have your thoughts about the best way to implement Dropout in blocks. In my opinion there are two options:
- Suggested by @bartvm: After construction of the model, manipulate the theano graph. One problem I see is that hidden units should be dropped, which, as far as I know, are not directly accessible with the Computational Graph. We can probably overcome this by cleverly masking weights and biases. 
- Make dropout part of the model architecture. We then need to have bricks that have a different apply methods at training time and test time.

**Comment by rizar** [comment]
Nice, I was just thinking about implementing it. 

As you said, there are two options. Personally I find the first better. Nice thing about Blocks is that hidden units _are_ accessible in the `ComputationGraph` by using the `VariableFilter`. Check [this line](https://github.com/bartvm/blocks/blob/master/examples/reverse_words/__init__.py#L187) for an example of retrieving activations of hidden units from a computation graph. This is possible because everything it tagged nicely behind the scenes when you call an application method.

There are caveats as well. The main trouble is that it is not that easy to insert something in an inner graph of a Scan Op. Especially hard should be to insert random number generation there. When a scan inner function uses randomness, `theano.scan` returns certain updates to be done, which makes me think that adding dropout inside a scan should be also accompanied by adding some updates, and to be honest I did not investigate this issue further. Finally there is a lot of confusion when one starts to think what the roles of new variables should be #175. 

The second option on the surface seems not to suffer from such problems, but it is effectively impossible to add a `Dropout` brick inside a ready model. However I think, we could support adding some pre- and post- processing to application methods like this:
```python
mlp = MLP([784, 500, 10], activations=[Tanh(), Tanh()])
mlp.linear_transformations[0].apply.after = dropout(0.1)
```
But this is a change to the model, and as you said the model is no longer suitable for usage, e.g. classification. But we could easily delete all this pre- and post- processing from the hierarchy. So this option can be considered as well.
f
To sum up: dropout as a graph transformation can be implemented very quickly in the similar way as `add_noise` is currently implemented. If one does not care about roles of "dropped out" variables and does not want to use it inside the scan, this will be a perfect solution for him. For more complicated cases more thinking is required.

**Comment by bartvm** [comment]
My preference is for the former approach, because it (theoretically) makes for nice and clean code, something along the lines of: `apply_dropout(cg, VariableFilter(roles=[INPUT])(cg.variables))` would work.

The naming scheme is a bit of rabbit hole, but in the end there I think there are very workable solutions, even if they're not as clean as we would like. A simple one would be this: Each replacement variable copies the roles but also has a tag (`.replacement_of = original_variable`). The `ComputationGraph` should then hardcode the behavior of not returning any variables which have replacements.

**Comment by dwf** [comment]
One thing to be certain of is that you want to also apply scaling by
inverse include probability at training time, so that you can use the
untransformed net at test time.

You also want to monitor most quantities using the untransformed graph, not
sure if that complicates things much.

**Comment by vdumoulin** [comment]
Are you guys talking about patching an existing computation graph to include dropout, or are you talking about creating _another_ computational graph with the same structure as the original but with dropout applied at some places?

It would make more sense to me to maintain two computational graphs, partly because as @dwf said monitoring usually include quantities computed on the untransformed graph.

I would imagine an interface that looks like that:

```python
from theano import tensor

from blocks.bricks import MLP, Rectifier, Softmax
from blocks.graph import ComputationGraph, apply_dropout


x = tensor.matrix('features')
y = tensor.matrix('targets')
mlp = MLP(activations=[Rectifier(), Softmax()], dims=[784, 500, 10])
mlp.initialize()
y_hat = mlp.apply(x)
cg = ComputationGraph([y_hat])
list_of_dropped_out_variables = # some filtering to get the desired variables
dropout_y_hat = apply_dropout(
    cg,  list_of_dropped_out_variables, include_prob=0.5, scale=2.0)
```

**Comment by bartvm** [comment]
Yes, that's exactly the plan! Internally we use Theano's [`clone` function](https://github.com/Theano/Theano/blob/master/theano/scan_module/scan_utils.py#L148-L171), so the inputs and shared variables should stay the same, but otherwise you should get a **copy** of the graph. This should make it quite straightforward to monitor variables in either of the two graphs.

The only issue holding this back right now is how we will tag the necessary variables (see lengthy discussion at #175). We could start with implementing the solution I just suggested (https://github.com/bartvm/blocks/issues/280#issuecomment-74341899), which should be fairly straightforward.

**Comment by rizar** [comment]
@bartvm , I like this new proposal of yours much more (in fact I also thought about it). We can have not only constant roles, but also carrying some information, e.g. a link to the variable that is replaced. This additional mark-up can be interpreted by the `VariableFilter`. We can have a hierarchy of such roles: e.g. a top role `ReplacementRole` and a role added by every graph transformation we have: `DropoutRole`, `NoisedRole` and so on.

So the principle would be: graph transformations never change roles of existing variables and never copy blindly roles from exiting variables to replacements, they assign new parametrized roles, and it is a task of filtering tools to interpret this rich markup.

**Comment by bartvm** [comment]
Mmm... Okay, that's actually slightly different from my idea, but I like it better. So here's how I imagine it in a bit more detail:

The replacement variable receives a tag, `tag.replaced = W1` which can be used by the filter. Ideally, I would like the `NOISED` roles to be dynamically created as to inherit from the role of the class they just replaced. That is, there should be a role `NOISED(WEIGHTS)`, `NOISED(BIASES)`, etc. where the class `NOISED(WEIGHTS)` inherits from `NoisedRole -> ReplacementRole`, but also `WeightsRole -> ParameterRole`.

In a case where only some of the weight matrices were noised, I'd imagine filtering to work like this:

```python
>>> VariableFilter(roles=[NOISED])(cg.variables)
[W1_noised]
>>> VariableFilter(roles=[NOISED])(cg.variables)[0].roles
[NOISED(WEIGHTS)]
>>> VariableFilter(roles=[WEIGHTS])(cg.variables)
[W1_noised, W1, W2]
>>> VariableFiler(roles=[WEIGHTS], exclude_replaced=True)(cg.variables)
[W1_noised, W2]
```

**Comment by harmdevries89** [comment]
Thanks @rizar, using the computational graph seems the best option. I have quickly implemented it as follows:
```python
def apply_dropout(cg, weights, hidden_states, p=0.5):
    srng = RandomStreams(seed=12345)
    dropout_states = zip(states, [s*srng.binomial(p=p, size=s.shape, dtype=floatX) 
                                  for s in hidden_states])
    dropout_weights = zip(weights, [w*(1/p) for w in weights])
     
    return cg.replace(dropout_states + dropout_weights)
```
The user is thus responsible for selecting the weights and hidden activities of the dropout layers. Does somebody have a elegant way to do that if you only want to apply dropout on a few layers? Right now, I gave activation layers a name with dropout in it (and used a regexp to select those activities), and selected the weights sort of manually.

**Comment by rizar** [comment]
@harmdevries89 , we have a `Selector` component for that, but it is not ready (there is a huge pull request #73 which is two month old and somebody should just clean it up and merge). The idea is to support selection of bricks by their paths from the root of the hierarchy, where path components can be regular expressions. Using this mechanism one should be able to select a subset of bricks. When such a subset is selected the respective hidden states can be retrieved by the `VariableFilter` (see [here](https://github.com/bartvm/blocks/blob/master/blocks/filter.py)).

That was a very high level description and is perhaps not too clear :) Btw, would you like to contribute this `apply_dropout` function?

**Comment by rizar** [comment]
@bartvm , I am not sure we need `NOISED(WEIGHTS)`. If we have a pointer to the replaced variable somewhere in the tag or in the role, the roles of the replaced variables are still accessible, so we do not need to duplicate information.

Also I would not use `tag.replaced` but have a `replaces` attribute for certain roles instead:

```python
class ReplacementRole(VariableRole):
   def __init__(self, replaces):
      self.replaces = replaces

class NoisedRole(VaribleRole):
   pass

>>> # Then we need filtering by class name
>>> VariableFilter(roles=[NoisedRole])(cg.variables)
[W1_noised]
>>> VariableFilter(roles=[NoisedRole])(cg.variables)[0].roles
[NoisedRole(W1_noised)]
>>> VariableFilter(roles=[WEIGHTS])(cg.variables) 
[W1, W2]
>>> VariableFilter(roles=[WEIGHTS], follow_links=True)(cg.variables) 
[W1_noised, W1, W2]
>>> VariableFiler(roles=[WEIGHTS], follow_links=True, exclude_replaced=True)
...     (cg.variables)
[W1_noised, W2]
```

**Comment by bartvm** [comment]
I don't like your defaults as much though. I think that if a user noises his weights, he should be able to retrieve them by default, not needing to know the meaning of an obscure `follow_links` argument. I'd like these replacements to be more transparent and easily accessible to users.

It's true that you can retrieve the original roles from the replaced variable, but creating the roles dynamically makes it possible for the user to interact with the roles directly more easily and do things like: `WEIGHTS in var.tag.roles` or `any(isinstance(_, PARAMETER) for _ in var.tag.roles)` to check whether a variable they have are weights or not, even if they are noised. Otherwise a user would have to immediately resort to writing stuff like
```python
def recursive_role_checker(var)
    if hasattr(var.tag, 'replaces'):
        recursive_role_checker(var.tag.replaces)
    else:
        return any(isinstance(_, PARAMETER) for _ in var.tag.roles)
```
Getting the non-noised variables should be solved by adding a `not_roles` keyword argument, which we should support anyway for cases like "give me all parameters except biases", as in:
```python
>>> VariableFilter(roles=[WEIGHTS], not_roles=[NOISED])(cg.variables)
[W1_noised, W2]
```
I think that being able to pass `not_roles` and being able to still interact with a noise weight matrix as if it is a weight matrix is more user-friendly, even if it adds a tiny bit of redundancy by wrapping the original role into a new one.

**Comment by rizar** [comment]
Not only adds it a tiny bit of redundancy, but also adds issues with serialization (as we are going to switch to cPickle we should mind it). How are you going to dynamically define roles in the global namespace?

**Comment by bartvm** [comment]
Instead of a factory pattern you could write a metaclass which adds the variables to the MRO, should be quite straightforward and I think it should pickle.

**Comment by dmitriy-serdyuk** [comment]
Have you thought about Boolean logic for roles? In order to be able to write: `my_role = WEIGHTS.and(NOISED.not)`. And have something like `check` function to check if a variable correspond to the equation (`my_role.check(x)`).

**Comment by bartvm** [comment]
Closed by #407
